Remove deprecated imsave method in neural style transfer example

stevemurr opened the pull request:
While attempting to run the neural style transfer example, the imsave method failed.  According to https://docs.scipy.org/doc/scipy/reference/generated/scipy.misc.imsave.html#scipy.misc.imsave  
`scipy.misc.imsave` is deprecated and `imageio.imwrite` should be used.  After making this change the script runs successfully.

Dref360 commented:
This adds a dependency?

stevemurr commented:
Yes, this adds `imageio` as a dependency.

fchollet commented:
Does scipy no longer have a built-in way to read/write images?

fchollet commented:
- Note that the save methods (`imread/imwrite`) are used in several other Keras examples
- In the `keras.preprocessing` module, we rely on PIL to open/save images. In order to avoid new dependencies, maybe we should use that here instead?

stevemurr commented:
Sounds good.  What do you think about adding the following function to `keras.preprocessing.image`  

```python
def save_img(path, x):
    """Save an image stored as a Numpy Array to a path or file object.

    # Arguments
        path: Path or file object
        x: Input Numpy Array
    """
    img = pil_image.fromarray(x)
    img.save(path)
```

fchollet commented:
Sounds reasonable. @Dref360 what do you think?

Dref360 commented:
We already have `array_to_img` which would make this feature more UX-friendly.
https://github.com/keras-team/keras/blob/master/keras/preprocessing/image.py#L277

save_img should have a similar API
Use : https://github.com/keras-team/keras/blob/master/keras/preprocessing/image.py#L1446

stevemurr commented:
Hi,

@Dref360 , I updated the `save_img` method to use `array_to_img` and updated the neural style transfer example to reflect the usage (I pushed an update to this PR that you can check out).  I also tested the change and verified images were saved as expected.  

@fchollet mentioned that numerous examples used the original `imread` and `imwrite` methods so if the `save_img` method looks good I can update and test the other example scripts that made use of the original `scipy.misc.imread` and `scipy.misc.imwrite` methods and update this PR. 

Let me know what you think.

Dref360 commented on examples/neural_style_transfer.py:
```diff
@@ -50,8 +50,7 @@
 from __future__ import print_function
-from keras.preprocessing.image import load_img, img_to_array
-from scipy.misc import imsave
+from keras.preprocessing.image import load_img, save_img, img_to_array
```
img_to_array is not used

stevemurr replied:
In this script `img_to_array` is used in the `preprocess_image` method here:
https://github.com/keras-team/keras/blob/b09ec1c9bf827622a5eebb94ba0af5bf202ca359/examples/neural_style_transfer.py#L98-L103

Dref360 commented on keras/preprocessing/image.py:
```diff
@@ -358,6 +358,17 @@ def img_to_array(img, data_format=None):
 
 
+def save_img(path, x):
+    """Save an image stored as a Numpy Array to a path or file object.
```
As I said, using img_to_array would be nicer. The API of `save_img` should reflect the one of img_to_array

stevemurr replied:
I forgot to add the change to my last commit - sorry about that.  The `save_img` now reflects the `array_to_img` API.

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -358,6 +358,19 @@ def img_to_array(img, data_format=None):
+    # Arguments
+        path: Path or file object.
+        x: Numpy array.
+        data_format: Image data format.
```
Explicitly mention that this is 'channels_first'/'channels_last' (since it may be confusing vs. image file format).

stevemurr replied:
I updated the docstring for the `data_format` parameter to explicitly state the options of `{"channels_first" or "channels_last"}`.

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -358,6 +358,19 @@ def img_to_array(img, data_format=None):
 
 
+def save_img(path, x, data_format=None, scale=True):
+    """Save an image stored as a Numpy Array to a path or file object.
```
Nit: "array" (uncapitalized)

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -358,6 +358,19 @@ def img_to_array(img, data_format=None):
     return x
 
 
+def save_img(path, x, data_format=None, scale=True):
```
I think we should add the arguments "file_format" and "**kwargs" (passed to `Image.save()`) https://pillow.readthedocs.io/en/3.1.x/reference/Image.html#PIL.Image.Image.save

"format" needs to be renamed "file_format" to avoid confusion with "data_format".

stevemurr replied:
Added `file_format` and `**kwargs` parameters which are passed to the `Image.save` method.

fchollet commented:
PEP8 error: https://travis-ci.org/keras-team/keras/jobs/371795626

stevemurr commented:
Cool!  I'll check the pep8 error, update the other example scripts and update this PR.

stevemurr commented:
Hi @fchollet , my PR fails a test regarding the undocumented `**kwargs` parameter passed to `save_img`.  I looked around the codebase and found instances of `**kwargs` that were not documented in their methods docstring.  What is the preferred pattern for documenting `**kwargs`?  I'm thinking I can either 1) Document the `**kwargs` in the docstring (which causes the test to pass) or 2) Disambiguate and document the `**kwargs` parameters passed to `Image.save`.  

Let me know what you think.

fchollet commented on examples/deep_dream.py:
```diff
@@ -83,7 +84,8 @@ def deprocess_image(x):
     # Add the L2 norm of the features of a layer to the loss.
-    assert layer_name in layer_dict.keys(), 'Layer ' + layer_name + ' not found in model.'
+    assert layer_name in layer_dict.keys(), 'Layer ' + layer_name + \
+        ' not found in model.'
```
Don't include this unrelated change

fchollet commented on examples/deep_dream.py:
```diff
@@ -67,6 +67,7 @@ def deprocess_image(x):
     x = np.clip(x, 0, 255).astype('uint8')
     return x
 
+
```
Don't include this unrelated change

fchollet commented on examples/neural_doodle.py:
```diff
@@ -347,6 +348,7 @@ def grads(self, x):
         self.grad_values = None
         return grad_values
 
+
```
Don't include these unrelated whitespace changes

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -358,6 +358,28 @@ def img_to_array(img, data_format=None):
+            format to use is determined from the filename extension.
+            If a file object was used instead of a filename, this
+            parameter should always be used.
+        scale: Whether to rescale image values to be within [0, 255].
```
Please add the line: "**kwargs: Additional keywords arguments passed to `PIL.Image.save()`". This will fix the CI error and it will improve the documentation for users.

stevemurr commented:
Hi, there seems to be a timeout issue in two of the travis build jobs that's effecting the recent PR's (mine included).  I noticed #10082 is addressing this.  I'll wait until that's resolved then trigger another build.
